### ISSUE TITLE
Add a code snippet to a result

### ISSUE THREAD

**Issue by michaelcfanning** [sarif-standard/sarif-spec#121]
is this simply a property? should we document common properties that producers add but which aren't specified? category and snippet are examples.

**Comment by ghost** [comment]
@michaelcfanning we've discussed this previously in the form of the question "should we define a reserved namespace of property bag property names?" (_e.g._, `sarif-category`, `sarif-snippet`). Our conclusion was that if something was important enough to reserve a property bag property name for, then it was important enough to represent as a first-class property of the property bag's parent (the `result`, or the `runInfo`, or whatever.

`category` in particular has come up before (#8), and we decided to add it to the `ruleDescriptor` object (but we never did so).

Jeffrey van Gogh wanted both `category` and `subCategory` which seems reasonable.

If we did that, we could make them multi-valued, which would reduce the pressure to allow multi-valued properties (although we might still want to do that). However, there are two issues:
1. Do we define a recommended set of values for `category`? If so, how strong is the recommendation? Or is it a free-for-all?
2. If we allow `category` to be multi-valued, and `subCategory` also to be multi-valued, how do we know which `category` each of the `subCategories are related to? We have a couple of options:
   
   a. Don't worry about that, just let it happen.
   
   b. Make `category` an object such as:

```
public class Category
{
    public string Name { get; set; }

    public List<string> SubCategories { get; set; }
```

**Comment by michaelcfanning** [comment]
Take another read of Jeff's mail, in which he likens subcategory to a rule id. I think you've got the right object definition for Category but, as usual would prefer not to design for complexity until we encounter/understand the scenarios that require it.

I don't think we can make a recommendation for category. In my experience, it is a dumping ground for annotations required to drive specific UX/automation scenarios (and is typically poorly maintained and sometimes riddled with inconsistences which are costly to fix later and aren't).

I have a specific idea I could offer here, but before that, can you tell me what distinguishes a category, exactly, from the 'tags' property?

**Comment by ghost** [comment]
I'm happy with the notion that `ruleId` is a good stand-in for `subcategory` so we don't need the latter.

I'm happy with allowing category values to be arbitrary strings.

To your question: since you mentioned "`tags`" rather than "`properties`", are you suggesting that users could write:

```
ruleInfo: [
  {
    ruleId: "CA1000",
    tags: ["security", "globalization", ...]
  }
]
```

... which would allow multiples without requiring multi-value properties?

They could also write `tags: ["category-security", "category-globalization", ...]`, although you might balk at the mini-language.

To answer your question: I'm fine with that.

(Fun fact: `ruleDescriptor` for some reason has `properties` but not `tags`. Everywhere else, they go together. I will add `ruleDescriptor.tags`, whether of not we do this.)

But to repeat my answer to your original question: we could add an Annex (non-normative) suggesting common property and tag values. But we shouldn't normatively suggest property bag property names, for the reason we've concluded in the past: any concept that deserves a normative property name deserves to be a first-class property.

**Comment by ghost** [comment]
Conclusion:
1. `tags` is close enough to `category` that we don't need to surface `category` as a first-class concept. Even if we did, we'd have trouble providing guidance on what to put in `tags` and what in `categories`, and viewers would have to decide how to treat both. So no, don't add `ruleDescriptor.categories`.
2. `codeSnippet` has the virtue of helping users debug a `result` entirely from data contained in the log file. So, yes, add `result.codeSnippet`.

**Comment by ghost** [comment]
Fixed in 03cb63a.

@michaelcfanning @nguerrera
